File: 1028 Recover a Tree From Preorder Traversal/1028.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode recoverFromPreorder(String S) {
        int level, nodeVal;
        Deque<TreeNode> stack = new LinkedList<>();
        // do not have i++
        for (int i = 0; i < S.length(); ) {
            for (level = 0; i < S.length() && S.charAt(i) == '-'; ++i) {
                level++;
            }

            for (nodeVal = 0; i < S.length() && S.charAt(i) != '-'; ++i) {
                nodeVal = nodeVal * 10 + (S.charAt(i) - '0');
            }
            while (level < stack.size()) {
                stack.pollFirst();
            }
            TreeNode newNode = new TreeNode(nodeVal);
            if (!stack.isEmpty()) {
                TreeNode tempNode = stack.peekFirst();
                if (tempNode.left == null) {
                    tempNode.left = newNode;
                } else if (tempNode.right == null) {
                    tempNode.right = newNode;
                }
            }
            stack.offerFirst(newNode);
        }
        while (stack.size() > 1) {
            stack.pollFirst();
        }
        return stack.pollFirst();
        
        
    }
    // Time Complexity: O(S)
    // Space Complexity: O(N)
}

File: 1028 Recover a Tree From Preorder Traversal/1028.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

import re

class Solution(object):
    def recoverFromPreorder(self, S):
        nodes = [(len(s[0]) ,int(s[1])) 
                for s in re.findall('(-*)(\d+)', S)[::-1]]
        def foo(level):
            if not nodes or level != nodes[-1][0]:
                return
            node = TreeNode(nodes.pop()[1])
            node.left = foo(level + 1)
            node.right = foo(level + 1)
            return node
        return foo(0)
            